khal/utils.py

```diff
@@ -590,19 +593,73 @@ def new_event(locale, dtstart=None, dtend=None, summary=None, timezone=None,
     return event
 
 
-def ics_from_list(vevent, random_uid=False):
-    """convert an iterable of icalendar.Event to an icalendar.Calendar
+def split_ics(ics):
+    """split an ics string into several according to VEVENT's UIDs
 
-    :param random_uid: asign the same random UID to all events
-    :type random_uid: bool
+    and sort the right VTIMEZONEs accordingly
+    ignores all other ics components
+    :type ics: str
+    :rtype list:
+    """
+    cal = icalendar.Calendar.from_ical(ics)
+    vevents = [item for item in cal.walk() if item.name == 'VEVENT']
+    tzs = {item['TZID']: item for item in cal.walk() if item.name == 'VTIMEZONE'}
+
+    events_grouped = defaultdict(list)
+    for event in vevents:
+        events_grouped[event['UID']].append(event)
+    return [ics_from_list(events, tzs) for uid, events in sorted(events_grouped.items())]
+
+
+def ics_from_list(events, tzs):
+    """convert an iterable of icalendar.Events to an icalendar.Calendar
+
+    :params events: list of events all with the same uid
+    :type events: list(icalendar.cal.Event)
+    :param tzs: collection of timezones
+    :type tzs: dict(icalendar.cal.Vtimzone
     """
     calendar = icalendar.Calendar()
     calendar.add('version', '2.0')
     calendar.add('prodid', '-//CALENDARSERVER.ORG//NONSGML Version 1//EN')
-    if random_uid:
-        new_uid = icalendar.vText(generate_random_uid())
-    for sub_event in vevent:
-        if random_uid:
-            sub_event['uid'] = new_uid
+
+    needed_tz, missing_tz = set(), set()
+    for sub_event in events:
+        # icalendar round-trip converts `TZID=a b` to `TZID="a b"` investigate, file bug XXX
+        for prop in ['DTSTART', 'DTEND', 'DUE', 'EXDATE', 'RDATE', 'RECURRENCE-ID', 'DUE']:
+            if isinstance(sub_event.get(prop), list):
+                items = sub_event.get(prop)
+            else:
+                items = [sub_event.get(prop)]
+
+            for item in items:
+                if not (hasattr(item, 'dt') or hasattr(item, 'dts')):
+                    continue
+                # if prop is a list, all items have the same parameters
+                datetime_ = item.dts[0].dt if hasattr(item, 'dts') else item.dt
+
+                if not hasattr(datetime_, 'tzinfo'):
+                    continue
+
+                # check for datetimes' timezones which are not understood by
+                # icalendar
+                if datetime_.tzinfo is None and 'TZID' in item.params and \
+                        item.params['TZID'] not in missing_tz:
+                    logger.warn(
+                        'Cannot find timezone `{}` in .ics file, using default timezone. '
```

Comment:
It might be useful as a separate command, like `repair`, which might eventually have other features as well.

I agree that it's best left outside the scope of this issue (so we can close this bug faster).